File: fragment/nodes/canvas.py
from .node import Node
from ..types.vector import Vector2
from pygame_render import Layer


class Canvas(Node):
    """The node responsible for rendering drawables.

    Canvas will render all drawable descendants captured by its cameras.
    """
    def __init__(self, *args, is_global_canvas: bool = False, **kwargs):
        super().__init__(*args, **kwargs)
        self.drawables = {} # dict([drawable1, drawable2...])
        self.cameras = []

        self.render_layer = self.game_manager.window_manager.renderer.make_layer((int(self.size[0]), int(self.size[1])))

        if not is_global_canvas:
            self.game_manager.window_manager.register_canvas(self)

    @property
    def size(self) -> Vector2:
        return self.game_manager.window_manager.window_size

    def register_camera(self, camera) -> None:
        self.cameras.append(camera)

    def unregister_camera(self, camera) -> None:
        self.cameras.remove(camera)

    def register_drawable(self, drawable) -> None:
        if self.drawables.get(drawable.draw_priority) is None:
            self.drawables[drawable.draw_priority] = []
        self.drawables[drawable.draw_priority].append(drawable)

    def unregister_drawable(self, drawable) -> None:
        self.drawables[drawable.draw_priority].remove(drawable)
        if len(self.drawables[drawable.draw_priority]) == 0:
            self.drawables.pop(drawable.draw_priority)

    def sort_drawable_priorities(self) -> None:
        """Sorts drawables according to their draw priorities."""
        sorted_drawables = {}
        for priority in sorted(self.drawables):
            sorted_drawables[priority] = self.drawables[priority]
        self.drawables = sorted_drawables

    def render(self) -> Layer:
        """Renders drawables captured by its cameras."""
        self.sort_drawable_priorities()
        self.render_layer.clear(0, 0, 0, 0)
        for camera in self.cameras:
            layer = camera.render()
            self.game_manager.window_manager.renderer.render(layer.texture, self.render_layer, (0, 0), camera.zoom)

        return self.render_layer
    
    def update_size(self) -> None:
        """Updates the render layer size according to the window size."""
        self.render_layer = self.game_manager.window_manager.renderer.make_layer((int(self.size[0]), int(self.size[1])))
    
    @classmethod
    def view_mode(cls, *args, **kwargs):
        return Node(*args, **kwargs)

File: fragment/nodes/canvas.js
import { Node } from '/fragment/nodes/node.js';

export class Canvas extends Node {
    /**
     * Canvas will "render" all drawable descendants captured by its active camera.
     */
    constructor(gameManager, properties, uuid = null, parent = null, scene = null, isGlobalCanvas = false) {
        super(gameManager, properties, uuid, parent, scene);
        this.activeCamera = null;
        this.registeredNodes = [];

        this.isGlobalCanvas = isGlobalCanvas;

        if (!this.isGlobalCanvas) {
            this.gameManager.windowManager.registerCanvas(this);
        }
    }

    get size() {
        return this.gameManager.windowManager.windowSize;
    }

    registerCamera(camera) {
        if (this.activeCamera === camera) {
            throw new Error('This camera is already registered as the active camera on this canvas.');
        }
        if (this.activeCamera !== null) {
            this.unregisterActiveCamera();
        }
        this.activeCamera = camera;
        this.gameManager.windowManager.app.stage.addChild(camera.pixiContainer);
        // Add previously unadded children
        this.registeredNodes.forEach(node => this.activeCamera.pixiContainer.addChild(node.pixiSprite));
    }

    unregisterActiveCamera() {
        if (this.activeCamera === null) {
            throw new Error('No active camera to unregister from this canvas.');
        }
        this.activeCamera.pixiContainer.removeChildren();
        this.gameManager.windowManager.app.stage.removeChild(this.activeCamera.pixiContainer);
        this.activeCamera = null;
    }

    registerDrawable(drawable) {
        if (this.registeredNodes.includes(drawable)) {
            throw new Error('Drawable already registered on this canvas.');
        }
        this.registeredNodes.push(drawable);
        if (this.activeCamera) {
            this.activeCamera.pixiContainer.addChild(drawable.pixiSprite);
        }
    }

    unregisterDrawable(drawable) {
        const index = this.registeredNodes.indexOf(drawable);
        if (index === -1) {
            throw new Error('Drawable not registered on this canvas.');
        }
        this.registeredNodes.splice(index, 1);
        if (this.activeCamera) {
            this.activeCamera.pixiContainer.removeChild(drawable.pixiSprite);
        }
    }

    destroySelf() {
        if (!this.isGlobalCanvas) {
            this.gameManager.windowManager.unregisterCanvas(this);
        }
        if (this.activeCamera !== null) {
            this.unregisterActiveCamera();
        }
        this.registeredNodes = [];
        super.destroySelf();
    }
}
